### PR TITLE
feat(controller): Add retry on different host

### PR DESCRIPTION
**Pull request by goock**
Fixes https://github.com/argoproj/argo/issues/4085

Issue requested `scheduleOnDifferentHostNodes` to be `boolean`, but due to different Kubernetes node's labels:
```
      retryStrategy:
        limit: 1
        scheduleOnDifferentHostNodes: true
```

I changed this to `ScheduleOnDifferentHostNodesLabel` as `string`:
```
      retryStrategy:
        limit: 1
        ScheduleOnDifferentHostNodesLabel: "kubernetes.io/hostname"
```
This PR introduces changes into:
- UI - add a list of failed hosts to retry node
- Retry strategy - new field `ScheduleOnDifferentHostNodesLabel`
- Controller

Checklist:

* [x] `make test` passed
* [x] ran `make codegen`
* [x] added documentation to `docs/examples/README.md`
* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

fail-hosts
fail-host2

### CODE REVIEW

File: ui/src/app/workflows/components/workflow-node-info/workflow-node-info.tsx
Comment (goock) [inline ui/src/app/workflows/components/workflow-node-info/workflow-node-info.tsx:99]: Because of this line UI test `Ensure Nothing Changed` failed.

Comment (goock) [reply ui/src/app/workflows/components/workflow-node-info/workflow-node-info.tsx]: I added missing tab.

**Comment by alexec** [comment]
Commentary:

* This is a great idea and we'd love to use it ourselves and we can see others using it too.
* We can also see other patterns, e.g. try with more memory. How could we generalize this? Could you introduce a new interface `ApplyRetryTweak(tmpl wfv1.Template, s wfv1.RetryStrategy, nodes wfv1.Nodes, node wfv1.NodeStatus)` that way we could isolate the code, it is easier to test, and extensible.
* I think all failed nodes should be available in `status/nodes` and you should be able to search that to identify failed hosts, no  need for extra fields.

**Comment by jessesuen** [comment]
Instead of:

```yaml
      retryStrategy:
        scheduleOnDifferentHostNodes: true
```

I think we'll need the ability to control affinity better (e.g. preferred vs. required) Could we instead do something like:

```yaml
      retryStrategy:
        hostAntiAfinity:
          requiredDuringSchedulingIgnoredDuringExecution: {}
```

```yaml
      retryStrategy:
        hostAntiAfinity:
          preferredDuringSchedulingIgnoredDuringExecution:
            weight: 1 # Between 1 - 100
```

**Comment by goock** [comment]
The idea behind this feature is to disallow to run the pod on the same host. We have thousands of hosts packed with many GPUs. Sometimes there is a software bug or just HW fails. We would love to have an option to retry the same template on a completely different host. This really would save resources.

This is my first PR for this project and I do not see the whole picture yet but let me left my comments:

- Generalize this feature would be a great idea but I think challenging. From my backyard, as an example, we would need feedback from the process what kind of error application generates, if this is OOM (CPU, GPU) or a different kind of bug
- Indeed, we can remove new field `failHostNodeNames` and loop over `status/nodes`.
- Should `ApplyRetryTweak()` be called from `executeTemplate()` as I did here https://github.com/argoproj/argo/pull/4679/files#diff-f321d4af83fcf8311dc80c0d50c59ac4c240f761206e7bb652709870eb9feb43R1682 or somewhere else?
- Option `preferredDuringSchedulingIgnoredDuringExecution` could lead back to the same host.
- Will `hostAntiAfinity`require to build completely new interface? I tried to use `StrategicMergePatch()` to merge nicely `Affinity` but it did not work. Eventually I finished with function `AddHostnamesToAffinity` https://github.com/argoproj/argo/pull/4679/files#diff-93ef535a58ca9268a8320d323c962ae9fdf2d3de0a54aa6e31f7fb652b5069c4R1056. With completly new struct `hostAntiAfinity` we will have to handle different kind of merge scenarios.

**Comment by alexec** [comment]
This is a very capable piece of work for a first PR. As step 1 - I'd ask you to check I'm right in thinking all the data you need is in `/status/nodes` and implement if that is the case. Lets deal with the generalized interface after that.

I think @jessesuen needs to answer the other Qs.

**Comment by goock** [comment]
I'll verify if all necessary data is in `status/nodes` and use it in the controller and ui

**Comment by goock** [comment]
@alexec @jessesuen I removed `failHostNodeNames` field from the controller and UI. It makes the entire process of rebasing much easier.

File: workflow/controller/operator.go
Comment (alexec) [inline workflow/controller/operator.go:2605]: (a) this is generic and so I don't think it should a utility func in the already bloated and hard to test woc.

(b) so I think this could be refactored as follows:

```golang
// Filter returns the children of the parent. 
func (s Nodes) Children(parentNodeId string) Nodes {...}
```


```golang
// Filter returns the subset of the nodes that match the predicate, e.g. only failed nodes
func (s Nodes) Filter(predicate func (NodeStatus) bool) Nodes {...}
```

```golang
// Map maps the nodes to new values, e.g. `x.Hostname`
func (s Nodes) Map(func (x NodeStatus) interface{}) map[string]interface{} {...}
```

```golang
var hostNames []string
for _, hostName := range woc.wf.Status.Nodes.
  Children(parent).
  Filter(func(x NodeStatus) {return x.Phase == NodeFailed || x.Phase == NodeError}).
  Map(func (x NodeStatus) {return x.Hostname)) {
    hostNames = append(hostNames, hostName.(string)
}
```

File: workflow/util/util.go
Comment (alexec) [inline workflow/util/util.go:1042]: please don't add extra functions to this already too large file, please add to a new file

File: workflow/controller/operator_test.go
Comment (alexec) [inline workflow/controller/operator_test.go:5493]: great test!

File: workflow/controller/operator_test.go
Comment (alexec) [inline workflow/controller/operator_test.go:5495]: can never be nil - assert redundant

File: workflow/controller/operator_test.go
Comment (alexec) [inline workflow/controller/operator_test.go:5497]: as above

File: workflow/controller/operator_test.go
Comment (alexec) [inline workflow/controller/operator_test.go:5499]: as above

File: workflow/controller/operator_test.go
Comment (alexec) [inline workflow/controller/operator_test.go:5501]: didn't know about this assert - could also use `assert.Empty`

File: workflow/util/util.go
Comment (alexec) [inline workflow/util/util.go:1056]: as above

File: workflow/controller/operator.go
Comment (alexec) [inline workflow/controller/operator.go:2609]: please don't panic - return nil or error instead

File: workflow/controller/operator.go
Comment (alexec) [inline workflow/controller/operator.go:1683]: so `executeTemplate` is the point in the code where we apply "retry tweaks"

File: workflow/controller/operator.go
Comment (alexec) [inline workflow/controller/operator.go:1776]: * move to own file
* this need not/should not be a receiver func (i.e. remove `woc`)
* remove return value - you already modify the param so this is redundant

create a new interface, something like

```golang
type RetryTweak = func (retryStrategy RetryStrategy, previousAttempts wfv1.Nodes, tmpl *wfv1.Template)
```

this func can then implement that interface

**Review by alexec — COMMENTED** [review]
This is looking fantastic. I'll keep trying to feedback quickly as you're making the changes. I've several comments...

File: pkg/apis/workflow/v1alpha1/workflow_types.go
Comment (goock) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1197]: I think this is the best place for those 3 functions.

File: workflow/controller/operator.go
Comment (goock) [reply workflow/controller/operator.go:1684]: I found this the best point for it, all the necessary data are here.

File: workflow/controller/retry_tweak.go
Comment (goock) [inline workflow/controller/retry_tweak.go:8]: @alexec I was not sure about the interface/implementation, I hope this is what you expected.

File: pkg/apis/workflow/v1alpha1/workflow_types.go
Comment (alexec) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:1197]: nice

File: workflow/controller/retry_tweak.go
Comment (alexec) [reply workflow/controller/retry_tweak.go:8]: right direction,  I'm thinking a function:

```golang
type RetryTweak = func(wfv1.RetryStrategy, wfv1.Nodes, *wfv1.Template)
```

File: workflow/controller/retry_tweak.go
Comment (alexec) [inline workflow/controller/retry_tweak.go:12]: with my idea, this would need to be a 2nd order function, e.g.

```golang
func RetryOnDifferentHost(retryHostName string) RetryTweak {
  return func(retryStrategy wfv1.RetryStrategy, allNodes wfv1.Nodes, tmpl *wfv1.Template) {
	hostNames := wfutil.GetFailHosts(allNodes, r.retryNodeName)
	hostLabel := retryStrategy.ScheduleOnDifferentHostNodesLabel
	if hostLabel != nil && len(hostNames) > 0 {
		tmpl.Affinity = wfutil.AddHostnamesToAffinity(*hostLabel, hostNames, tmpl.Affinity)
	}
}
```

File: workflow/util/retry.go
Comment (alexec) [inline workflow/util/retry.go:10]: maybe we could move this all to it's own package? workflow/util/retry?

File: pkg/apis/workflow/v1alpha1/workflow_types.go
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1341]: if empty string is not a valid value, then you can make this simpler using `string` instead of `*string`

File: pkg/apis/workflow/v1alpha1/generated.proto
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/generated.proto:865]: @jessesuen thoughts on the proto?

File: workflow/controller/operator.go
Comment (goock) [reply workflow/controller/operator.go]: I implemented this in a new file.

File: pkg/apis/workflow/v1alpha1/generated.proto
Comment (alexec) [reply pkg/apis/workflow/v1alpha1/generated.proto:865]: could we use something like a NodeSelector instead of string?

File: workflow/util/util.go
Comment (goock) [reply workflow/util/util.go]: Moved to a new file

File: workflow/controller/operator.go
Comment (goock) [reply workflow/controller/operator.go]: Panic is removed now.

File: workflow/util/retry.go
Comment (goock) [reply workflow/util/retry.go]: As you suggested, I moved `retry.go` and `retry_test.go` to `workflow/util/retry` directory.

File: pkg/apis/workflow/v1alpha1/workflow_types.go
Comment (goock) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: I think it should be `ID` here, it's changed.

File: workflow/controller/operator.go
Comment (goock) [inline workflow/controller/operator.go:1685]: The retry node has the same `Name` and `ID` so can we stick with `retryNodeName` or I should change and use `ID` here?

File: pkg/apis/workflow/v1alpha1/workflow_types.go
Comment (goock) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: That's changed too

File: pkg/apis/workflow/v1alpha1/generated.proto
Comment (goock) [reply pkg/apis/workflow/v1alpha1/generated.proto:865]: A label like `"k3s.io/hostname"` it's unique across all nodes in the cluster.
With `NodeSelector` we could use any label but this not give us uniqueness, moreover, this will complicate this PR. We have already to use the field `NodeStatus.HostNodeName`.

Comment (alexec) [reply pkg/apis/workflow/v1alpha1/generated.proto:865]: maybe

```yaml
retryStrategy:
    nodeAffinity:
      differentHost: true  
      nodeAffinity: ...
```

File: workflow/controller/retry_tweak.go
Comment (goock) [reply workflow/controller/retry_tweak.go:16]: That's done

File: pkg/apis/workflow/v1alpha1/generated.proto
Comment (alexec) [reply pkg/apis/workflow/v1alpha1/generated.proto:865]: @jessesuen bump.

**Comment by salanki** [comment]
@alexec @jessesuen: Can we get this merged with existing functionality, and add any more dynamic node affinity logic in a subsequent feature?

**Comment by alexec** [comment]
@salanki I've spoken to @jessesuen and this is our proposal - thoughts?

```yaml
retryStrategy:
  affinity: # to allow for podAntiAffinity in the future
    nodeAntiAffinity:
      labelSelector: kubernetes.io/hostname # if omitted, this is the default 
---
retryStrategy:
  affinity:
    nodeAntiAffinity: { }
```

**Comment by salanki** [comment]
Seems reasonable with room for expansions. To enable this feature, while using the default label selector, would you define:
```
retryStrategy:
  affinity: # to allow for podAntiAffinity in the future
    nodeAntiAffinity: {}
```

Or what is your thinking there?

**Comment by alexec** [comment]
@salanki - that's correct - you can omit `labelSelector` and get the default

**Comment by salanki** [comment]
Sounds reasonable. @goock will you update the PR for the new syntax please!

**Comment by goock** [comment]
@alexec
If I am not wrong, the change requires to retrieve labels from node. I'm thinking to add function like this:

```
func getNodeLabels(config *rest.Config, hostname string) (map[string]string, error) {
	clientset, err := kubernetes.NewForConfig(config)
	if err != nil {
		panic(err.Error())
	}

	labels := make(map[string]string)
	listOptions := metav1.ListOptions{
		LabelSelector: "kubernetes.io/hostname=" + hostname,
	}
	nodes, err := clientset.CoreV1().Nodes().List(listOptions)
	if err != nil {
		return labels, err
	}

	if len(nodes.Items) != 1 {
		return labels, nil
	}

	for label, value := range nodes.Items[0].ObjectMeta.Labels {
		labels[label] = value
	}

	return labels, nil
}
```
The `config` is available as `woc.controller.restConfig`.

Is `workflow/controller/controller.go` the right place to add this function?

**Comment by alexec** [comment]
Good point. This will be expensive to do. We should avoid it.

**Comment by alexec** [comment]
What if we remove the `labelSelector` today?

```yaml
retryStrategy:
  affinity:
    nodeAntiAffinity: {}
```

**Comment by goock** [comment]
> What if we remove the `labelSelector` today?
> 
> ```yaml
> retryStrategy:
>   affinity:
>     nodeAntiAffinity: {}
> ```

Ok, let's get rid of `labelSelector` for now and use the default label `kubernetes.io/hostname`.

I hope, according to this https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/#kubernetesiohostname
that all clusters will have the standard label `kubernetes.io/hostname`.

**Comment by alexec** [comment]
I don't think I was clear. We need to avoid any new API requests. So `getNodeLabels` isn't going to work. Instead, let's take `nodeAntiAffinity: {}` to imply the use don't want the retry to run on any of the previous failed nodes.

**Comment by goock** [comment]
@alexec Could we merge this PR? I implemented all of the requested features.

File: examples/README.md
Comment (alexec) [inline examples/README.md:816]: is this still correct?

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
Requires approval from @jessesuen

File: examples/README.md
Comment (goock) [reply examples/README.md:816]: yes, it uses the default selector for hostname `kubernetes.io/hostname`. It seems to be a k8s wide standard according to this: https://kubernetes.io/docs/reference/kubernetes-api/labels-annotations-taints/#kubernetesiohostname

**Review by jessesuen — APPROVED** [review]

**Comment by alexec** [comment]
You have full approval - but the conflicts need to be resolved before we can merge it.

**Comment by goock** [comment]
@alexec I just resolved the conflicts.
For sure this is my first contribution but the last one to this great project. Thank you guys for your patience and support during this PR.

**Comment by salanki** [comment]
Great work @goock!

**Comment by 3072141364** [comment]
I tried，but not work

**Pull request merged**
